hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/CaseOperationIntegrationTest.java

```diff
@@ -0,0 +1,362 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression;
+
+import com.hazelcast.sql.SqlColumnType;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.SqlErrorCode;
+import com.hazelcast.sql.impl.expression.predicate.ComparisonMode;
+import com.hazelcast.sql.impl.expression.predicate.ComparisonPredicate;
+import com.hazelcast.sql.support.expressions.ExpressionBiValue;
+import com.hazelcast.sql.support.expressions.ExpressionType;
+import com.hazelcast.sql.support.expressions.ExpressionTypes;
+import org.junit.Test;
+
+import java.io.Serializable;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.OffsetDateTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+
+import static com.hazelcast.sql.impl.type.QueryDataType.INT;
+import static com.hazelcast.sql.impl.type.QueryDataType.VARCHAR;
+
+public class CaseOperationIntegrationTest extends ExpressionTestSupport {
+    @Test
+    public void caseWithConstants() {
+        put(1);
+
+        checkValue0("select case when true then null else null end from map", SqlColumnType.NULL, null);
+        checkFailure0("select case when 1 then 1 else 2 end from map", SqlErrorCode.PARSING, "Expected a boolean type");
+        checkValue0("select case when 1 = 1 then 1 else null end from map", SqlColumnType.TINYINT, (byte) 1);
+        checkValue0("select case when 1 = 1 then null else 1 end from map", SqlColumnType.TINYINT, null);
+        checkValue0("select case 1 when 1 then 100 else 2 end from map", SqlColumnType.TINYINT, (byte) 100);
+        checkFailure0("select case 'a' when 1 then 100 else 2 end from map", SqlErrorCode.PARSING, "Cannot apply '=' operator to [VARCHAR, TINYINT]");
+    }
+
+    @Test
+    public void testOnlyElse() {
+        put(1);
+
+        String sql = "select case else 2 end from map";
+
+        checkFailure0(sql, SqlErrorCode.PARSING, "Encountered \"case else\"");
+    }
+
+    @Test
+    public void useMapValue() {
+        put(1);
+
+        checkValue0("select case when this = 1 then 10 end from map", SqlColumnType.TINYINT, (byte) 10);
+    }
+
+    @Test
+    public void multipleConditions() {
+        put(1);
+
+        String sql = "select case when this > 1 then 10 when this = 1 then 100 end from map";
+
+        checkValue0(sql, SqlColumnType.TINYINT, (byte) 100);
+    }
+
+    @Test
+    public void doesntMatchWhen_andNoElseBranch() {
+        put(1);
+
+        String sql = "select case when this > 1 then 10 end from map";
+
+        checkValue0(sql, SqlColumnType.TINYINT, null);
+    }
+
+    @Test
+    public void differentReturnTypes() {
```

Comment:
Would it be useful to enumerate and test all incompatible return types + all compatible return types (if implicit conversion can be made)?

Reply:
Sure I will add such tests, but what worries me is that each `@Test` method for case operator takes `> 2sec`. If we repeat the same tests for incompatible and compatible types for each operator test time will grove quickly.

I am not sure if we have "single" entry point for type coercion in query execution, realising that this PR has a copy-past code.
This https://github.com/hazelcast/hazelcast/pull/18067/files#diff-cad736f34f86bc13400d98e3ecda75f7333d19db3e5b32cbb8329ed2302649ceR165 was copied from https://github.com/hazelcast/hazelcast/blob/f783ed84d4624eb4556f5b9ab18d2247c68d8a66/hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/predicate/HazelcastComparisonPredicateUtils.java#L115